sessionmanager.md

```diff
@@ -0,0 +1,117 @@
+# `SessionManager` for @azure/service-bus v7
+
+## Introduction
+
+Sessions allow a user to logically subdivide a queue or subscription
+into multiple smaller queues. This is useful for customers that would 
+like to allow for many smaller streams of messages within a single 
+queue or subscription.
+
+## Use case
+
+A simple use case would be having a set of documents. Users can "edit"
+documents, and each edit is stored as a message within Service Bus.
+
+```javascript
+{ 
+  sessionId: "the document name"
+  body: {
+    mode: "edit",
+    content: "content of the edit"
+  }  
+}
+```
+
+Multiple edits or transformations could be sent for a document, at any time.
+
+To handle this the user would need to write code similar to this:
+
+```javascript
+
+  while (true) {    
+    // some sort of check to make sure we're not processing
+    // too many sessions at a time. The odds of having more documents
+    // than available processors is _high_.
+    acquireSemaphore();
+
+    // by not specifying a session id we will get the next unlocked session from service bus.
+    const sessionReceiver: SessionReceiver = serviceBusClient.getSessionReceiver("queue", "peekLock");
+    startSessionLockRenewal(sessionReceiver);
+
+    sessionReceiver.subscribe(messageHandlers);
+
+    // ...after some predefined period of time
+    sessionReceiver.close();
+
+    stopSessionLockRenewal(sessionReceiver);
+    releaseSemaphore();
+  }
+```
+
+## Proposal
+
+In the same vein as what we provide in `Receiver.subscribe()` we can provide a 
+`SessionManager` that would support the use case where the user would like 
+to read messages, in parallel, from a number of sessions while controlling
+the amount of parallelism.
+
+This would simplify session usage to the point where it becomes more accessible, 
+allowing more customers to use this powerful feature.
+
+The proposed interface for this change:
+
+```javascript
+class ServiceBusClient {
+  // a session manager is conceptually a receiver, but does not have to 
+  // obey the same interface.
+  createSessionManager("queue", "peekLock", sessionManagerOptions): SessionManager;
```

Comment:
Is it feasible to support the case where the user wants to manage _known_ sessions by allowing them to pass a set of `session-id`s ?

Reply:
I was thinking 'no' but really only because if it's a fixed or known set of sessions then you don't really need this class.

When you're in a "I know all the IDs" situation things are actually more similar to just having a queue or subscription. My assumption (possibly bad!) is that the number of sessions is going to be bounded and so not as difficult to manage. 

@JoshLove-msft  - in your current processor I believe (correct me, please) that you can basically do:
1. A single known session ID
2. Any # of unknown session IDs
3. No sessions, just a queue or subscription

Is that correct?

Reply:
Correct, we don't currently support specifying a list of named sessions.

Reply:
@HarshaNalluru  - what do you think? Are we missing out on an interesting use-case?

Reply:
> what do you think? Are we missing out on an interesting use-case?

Nothing interesting, but...
Say... a queue has 10 sessions and the user wants to receive from only 5 known sessions, instead of forcing the user to open 5 different SessionReceivers, this SessionManager could potentially provide a great convenience for such case.
(Maybe this case is more relevant once we start to allow `listing-sessions-in-an-entity` through our SDK.)

Reply:
For the .NET perspective, if there was a use case for a list of named sessions, we could add this in a non-breaking way as a new overload for CreateSessionProcessor.

Reply:
Funny but two days ago I asked the same exact question from @JoshLove-msft with exact same example( @HarshaNalluru you won't believe but I've used 10 and 5 numbers in my example)  :smiley:

Reply:
Funny but two days ago I asked the same exact question from @JoshLove-msft with exact same example(@HarshaNalluru you won't believe but I've used 10 and 5 numbers in my example)  :smiley:

Reply:
I think we can group this under "future enhancement", primarily because it appears we can offer this type of thing in the future compatibly.

I think if we keep this scoped to the one case (unbounded, unknown sessions) then we're in good shape. My primary goal with this is to make people aware of it and make it accessible.

I **think** (emphasis) handling multiple known receivers is a pretty simple thing to do and probably isn't nearly in the same ballpark of scale.

Reply:
I agree this can be added later on.

Reply:
Heh, as seems to be the case I'm on the other side of pretty much all these opinions.  I think the narrow scoping defeats most of the benefit of this, and that MANY of my session cases will be "only this subset" and not "all sessions."  I'm likely slightly colored by the fact that we allowed specified sessions in the existing python.

That said, as I mentioned to you in chat I'm more concerned about situations such as: 
1. What happens if someone runs two of these on two machines at once?  They just greedily take all available sessions?  
2. What happens if a lock is lost while processing a set of receivers on this machine, but another manager on another machine claims it?

Reply:
> I think the narrow scoping defeats most of the benefit of this, 

I haven't heard _why_ we think that is. I have one specific customer case that I'm building this for that I think showcases a great use case for sessions where they are unbounded and they want to round-robin them to process it. This PR would address their case perfectly, as far as I can see.

> and that MANY of my session cases will be "only this subset" and not "all sessions." 

The "only this subset" is very easily handled by the customer opening up SessionReceiver's. We can always take feedback later if that turns out to be a case people need to solve - ie, not all of the problems have to be solved today and I don't think we've precluded it from being solved in any of our designs.

> I'm likely slightly colored by the fact that we allowed specified sessions in the existing python.

Totally understand. :)

> 
> That said, as I mentioned to you in chat I'm more concerned about situations such as:
> 
>     1. What happens if someone runs two of these on two machines at once?  They just greedily take all available sessions?

The handing out of sessions happens service-side. When the user creates a `SessionManager` they have two dials they can adjust:

1. Number of sessions to concurrently receive from
2. Number of messages, per receiver, that can be in flight.

If they don't coordinate those they get their chaos they'd get if they had done the loop themselves.

> 
>     2. What happens if a lock is lost while processing a set of receivers on this machine, but another manager on another machine claims it?

This locking is handled service side and this case should be expected by clients. A session manager might "drop out" entirely because the client crashes, exits, etc...

This is pretty much the same case as peekLock - if a client stops renewing a lock it expires and someone else can pick it up afterwards.

Reply:
Based on our offline discussions I believe we've arrived on this design _at_ least working for sessions. Future expansion might happen but it should be possible. within the bounds of JS at this point.